File: recursion/python/solutions/sum_squares.py
"""Solution for sum_squares."""

from linked_list import EMPTY_LIST, Node, is_empty, prepend, unprepend


def sum_squares(lst):
    """
    Return the sum of squares of all elements.

    sum_squares(1 -> 2 -> 3) => 1 + 4 + 9 = 14
    sum_squares(()) => 0
    """
    if is_empty(lst):
        return 0

    first, rest = unprepend(lst)

    return first * first + sum_squares(rest)


if __name__ == '__main__':
    # Test cases
    assert sum_squares(Node.from_list([1, 2, 3])) == 14
    assert sum_squares(Node.from_list([1, 1, 1])) == 3
    assert sum_squares(Node.from_list([5])) == 25
    assert sum_squares(EMPTY_LIST) == 0
    assert sum_squares(Node.from_list([0, 0, 0])) == 0

    print("All tests passed!")

File: recursion/javascript/solutions/sumSquares.js
/**
 * Solution for sumSquares.
 */

const { EMPTY_LIST, Node, isEmpty, unprepend } = require('./linkedList');

function sumSquares(list) {
  if (isEmpty(list)) {
    return 0;
  }

  let [first, rest] = unprepend(list);

  return first * first + sumSquares(rest);
}

if (require.main === module) {
  console.assert(sumSquares(Node.fromArray([1, 2, 3])) === 14);
  console.assert(sumSquares(Node.fromArray([1, 2, 3, 4])) === 30);
  console.assert(sumSquares(Node.fromArray([5])) === 25);
  console.assert(sumSquares(EMPTY_LIST) === 0);

  console.log('All tests passed!');
}

module.exports = { sumSquares };
